// pretext-pdf/dist/builder.js
/**
 * Builder API for pretext-pdf
 *
 * Provides a fluent, chainable interface for constructing PDF documents.
 * Accumulates ContentElement[] and delegates final rendering to render().
 */
import { render } from './index.js';
/**
 * Create a new PDF document using the builder API.
 *
 * @example
 * ```ts
 * const pdf = await createPdf({ pageSize: 'Letter' })
 *   .addHeading('Hello')
 *   .addText('This is a paragraph.')
 *   .addHr()
 *   .build()
 * ```
 */
export function createPdf(options = {}) {
    const content = [];
    return {
        /**
         * Add a paragraph.
         */
        addText(text, opts) {
            content.push({ type: 'paragraph', text, ...opts });
            return this;
        },
        /**
         * Add a heading.
         */
        addHeading(text, opts) {
            const level = (opts?.level ?? 1);
            const { level: _, ...restOpts } = opts ?? {};
            content.push({ type: 'heading', level, text, ...restOpts });
            return this;
        },
        /**
         * Add a table.
         */
        addTable(opts) {
            content.push({ type: 'table', ...opts });
            return this;
        },
        /**
         * Add an image.
         */
        addImage(src, opts) {
            content.push({ type: 'image', src, ...opts });
            return this;
        },
        /**
         * Add a list (ordered or unordered).
         */
        addList(opts) {
            content.push({ type: 'list', ...opts });
            return this;
        },
        /**
         * Add a code block.
         */
        addCode(text, opts) {
            content.push({ type: 'code', text, ...opts });
            return this;
        },
        /**
         * Add a rich paragraph with mixed formatting (bold, italic, color).
         */
        addRichText(spans, opts) {
            content.push({ type: 'rich-paragraph', spans, ...opts });
            return this;
        },
        /**
         * Add a blockquote.
         */
        addBlockquote(text, opts) {
            content.push({ type: 'blockquote', text, ...opts });
            return this;
        },
        /**
         * Add a horizontal rule.
         */
        addHr(opts) {
            content.push({ type: 'hr', ...opts });
            return this;
        },
        /**
         * Add a spacer (vertical whitespace).
         */
        addSpacer(height) {
            content.push({ type: 'spacer', height });
            return this;
        },
        /**
         * Add a page break.
         */
        addPageBreak() {
            content.push({ type: 'page-break' });
            return this;
        },
        /**
         * Get the underlying declarative document.
         * Useful for inspection, serialization, or reusing with render().
         */
        toDocument() {
            return {
                pageSize: options.pageSize,
                margins: options.margins,
                defaultFont: options.defaultFont,
                defaultFontSize: options.defaultFontSize,
                defaultLineHeight: options.defaultLineHeight,
                fonts: options.fonts,
                header: options.header,
                footer: options.footer,
                metadata: options.metadata,
                content,
            };
        },
        /**
         * Build the PDF and return the bytes.
         * Delegates to render() with the accumulated document.
         */
        async build() {
            return render(this.toDocument());
        },
    };
}
